Caesar Cipher

whiletruelearn opened the pull request:
- [x] pep8
- [x] docstrings
- [x] unit tests
- [x]  API manual
- [x]  changelog

Adding <code>encrypt()</code> and <code>decrypt()</code> functionality for caesar cipher

Refer Issue #4

whiletruelearn committed:
Changes for adding encrypt and decrypt for caesar cipher

tests for caesar cipher

changelog updated with entry for caesar cipher

caesar with encrypt and decrypt function

tests for caesar cipher

Fixing PEP8 violations, Adding test cases for 100% line coverage

bravegnu commented:
@kskrishnasangeeth can you update the checklist in issue #4.

kskrishnasangeeth commented:
@bravegnu

[X] pep8
[X] docstrings
[X] unit tests
[ ]  API manual
[ ]  changelog

I am sorry. I didn't quite get the API manual part. Do you want me to update README with details on Caesar cipher or is it something else ?

I have updated the changelog but is a bit sketchy on whether it is correct format.

bravegnu commented:
The details are in https://github.com/Chennaipy/libcipher#contributing-guidelines Please let me know if that doesn't help.

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,64 @@
+def encrypt(message, key=None):
+    """Returns the encrypted messsage for the string passed as argument.
+
+    Uses a random key for generating the cipher text
```
Selecting a random key, that user itself will not know, does not seem to be useful. I don't think this is required.

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,64 @@
+
+    if key <= 0 or key >= 27:
+        print('Key should be having value between 1 and 26 ')
+        sys.exit(0)
```
You should never quit from a library function. Raise an exception to the caller, and let the caller decide what to do.

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,64 @@
+    return cipher_text
+
+
+def decrypt(message, key):
```
Both `encrypt()` and `decrypt()` are identical. Can we use a core private method that does both and implement the encrypt and decrypt as wrappers of it?

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,64 @@
+        key (int) : The key based on which encryption is performed. Is optional value.
+
+    Return:
+        str: the encrypted message """
```
Move the triple quote to a new line. That is the pep8 recommendation.

```
Unless the entire docstring fits on a line, place the closing quotes on a line by 
themselves. This way, Emacs' fill-paragraph command can be used on it. 
```

shrayasr replied:
> This way, Emacs' fill-paragraph command can be used on it. 

That's quite _editor-ist_ don't you think? :wink:

bravegnu replied:
How dare you disgrace Emacs by calling it an editor.

shrayasr replied:
How I imagined you saying that: 

![56647084](https://cloud.githubusercontent.com/assets/240368/7204995/d815d7e4-e543-11e4-8033-5015025f82fa.jpg)

bravegnu replied:
:smile:

whiletruelearn commented:
@bravegnu 

Thanks for the code review. I will work on these tonight . Sorry for the delay in updating . Held up with few other things .

whiletruelearn committed:
Raising custom exceptions instead of exit, Fixing PEP8 violations, removing random key gen logic, correcting docstrings and modularising the code to remove redundant code.

Updating index.rst and caesar.rst

Changes based on the diff comments

Delete invalid_key.py

Diff changes

Fix Exception testing fail for Python 2.x

whiletruelearn commented:
@bravegnu 

Incorporated the changes you suggested. Travis build is failing for <code>Python 2.6</code> and <code>2.7</code> in test cases which checks if a correct exception is being called. 

I have tested out on <code>Python 3.4</code> and it ran fine .  The issue IMO is because of the syntax difference on how exception is handled in 2.x and 3.x 
- 2.x

``` Python
   except Exception, err
```
- 3.x

``` Python
 except Exception as err
```

bravegnu commented on doc/caesar.rst:
```diff
@@ -0,0 +1,5 @@
+Reverse Cipher
```
This needs to be fixed. :smile:

bravegnu commented on doc/changelog.rst:
```diff
@@ -1,6 +1,6 @@
 Changelog
 =========
-
+* :support:`4` Add encryption and decryption for caesar cipher
```
Nitpick. Needs a full stop.

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,107 @@
+
+
+def cipher_helper(message, key):
+
```
Blank line not required.

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,107 @@
+    """Function for modularising and removing some of the repeated code.
+
+    Attributes:
+
```
Here too.

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,107 @@
+
+    Attributes:
+
+        message(str) -- Message that is supplied for encryption/decryption.
```
Sorry but you need to follow the convention too the word. A Sphinx plugin called Napoleon will generate documentation from this. See http://sphinxcontrib-napoleon.readthedocs.org/en/latest/example_google.html

whiletruelearn commented:
@bravegnu Made the changes based on diff comments.

shrayasr commented:
Build seems to be failing on Python 2.6 and 2.7 boss. Please check on those as well

shrayasr commented:
> I have tested out on Python 3.4 and it ran fine . The issue IMO is because of the syntax difference on how exception is handled in 2.x and 3.x
> 
> 2.x
> 
> ```
>    except Exception, err
> ```

You can handle exceptions in 2.x as `except Exception as err` as well :)

whiletruelearn committed:
Fix for test Exception fail in Python 2.6

Using modulo logic , fixing pep8 violations

whiletruelearn commented:
Thanks @shrayasr . 

I missed that. Fixing this has been fun. :smile:  Yay! The build passed.

It was working fine in 3.x which i have in my machine but failed for 2.x. Then i made some changes by using <code>with clause</code>  and it worked fine in 2.7 but not in 2.6. 

Found something in stackoverflow and that seems to get the build passed.

shrayasr commented:
Great :+1: 

Also if you look at the state of the PR right now, there are 13 commits which might not be a good idea for one functionality. My recommendation is to [Squash](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) these commits into a few that make sense and doing a `push --force` to update them :) 

Please note: It is a very _bad_ thing to do a `push --force` but its fine in this case since it is on a PR and you aren't changing **everyone**'s history. [Here](http://www.mail-archive.com/dri-devel@lists.sourceforge.net/msg39091.html) is an excerpt of the boss (Linus) himself talking about it :)

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,108 @@
+        key (int): The key based on which encryption is performed.
+
+    Returns:
+        cipher_text (str): the encrypted message.
```
That would be

```
str: the encrypted message
```

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,108 @@
+    return message.lower()
+
+
+def encrypt(message, key=None):
```
Why would you want to make it optional and later throw an error that user did not provide it?

whiletruelearn replied:
Key is mandatory. It is initialized as None for the unit test function

``` Python
def test_encrypt_no_key(self):
         self.assertRaises(NoKeyGiven, encrypt, "hello")
```

bravegnu replied:
Point 1, if you really want to trigger that exception,

```
self.assertRaises(NoKeyGiven, encrypt, "hello", None)
```

Point 2, do you really have to check for `key is None`. And why stop there and why not check `message is None`.

whiletruelearn replied:
Agree. I can make the changes so that test cases are not impacted.

whiletruelearn replied:
@bravegnu Few more thoughts on this.

The reason i made <code>key</code> initialized to _None_ at first place is when we do this 

<code>>>>encrypt("message") </code>
I get this as the error message.

``` Python
TypeError: encrypt() missing 1 required positional argument: 'key'
```

which is the default error message.

So my custom Exception message becomes pointless as we really don't do things like
<code>encrypt("hello",None)</code>  in real scenarios  to get the custom message below

``` Python
    raise NoKeyGiven("Key is needed to be entered")
```

So it leaves me with two options 
a) Remove the exception for missing Key
b) Leave the key as optional in <code>encrypt()</code> and <code>decrypt()</code> functions. 

What do you reckon is the best thing to do?

> Point 2, do you really have to check for key is None. And why stop there and why not check message is None.

This will happen only when we try something like this right

<code>>>>encrypt( ,10) </code>

This appears to be syntactically wrong and i guess we don't need to have an exception for it.

Or a scenario where message is missing could be <code>>>>encrypt(10)</code> , but i am not sure if this scenario is worth pursuing.

whiletruelearn replied:
On a lighter note

> Point 2

<img src="https://cloud.githubusercontent.com/assets/1977850/7327001/5f6ec554-eae7-11e4-8a92-5eae9808c691.jpg" </img>

shrayasr replied:
> Was that sarcasm?

![94f0cb21bd56b93e42b4e712200eced6](https://cloud.githubusercontent.com/assets/240368/7327778/77595674-eaed-11e4-9775-041899c77a40.jpg)

I think so :smile:

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,108 @@
+            changed_char = chr(ord(each_char) + key)
+            if ord(changed_char) > ord('z'):
+                changed_char = chr(ord(each_char) + (key - 26))
+            cipher_text += changed_char
```
@kskrishnasangeeth if this can be implemented using modulo arithmetic as shown below, it will be more elegant. What do you say?

```
    chpos = string.lowercase.find(each_char)
    if chpos == -1:
        cipher_text += each_char
        continue
    else:
        chpos = (chpos + key) % 26
        cipher_text += string.lowercase[chpos]
```

Also, because of the way modulo arithmetic works with negative numbers, decryption is all about changing the sign of the key. And rest of the code will be the same!

whiletruelearn replied:
Agree . This is more cleaner solution. I didn't know the modulo arithmetic property on negative number. Thanks for the suggestion.

whiletruelearn committed:
Using modulo logic, fixing pep8 violations , fixing build failure for 3.x

whiletruelearn commented:
@shrayasr Thanks ! Rebase really worked like a charm :sweat_smile:

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,112 @@
+
+    Returns:
+        str: the encrypted message.
+
```
Blank line before `"""` is not required. But after it is useful.

shrayasr replied:
@kskrishnasangeeth fix this as well

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,112 @@
+    return cipher_text
+
+
+def decrypt(message, key):
```
Both the `encrypt()` and `decrypt()` are nearly identical. This is a violation of the DRY rule. How about reducing the decrypt to:

```
encrypt(messge, -key)
```

whiletruelearn replied:
But what if for <code>encrypt()</code> , we supply negative key . Should that be supported as a valid case for <code>encrypt()</code>? Currently we are raising an exception if <code>key<=0</code>. 

If i have something like <code> decrypt_text = encrypt(message, -key) </code> , then the exception gets triggered and the existing test cases for <code>decrypt()</code> fails. 

Even if the code is repeated , IMHO it  is okay .

Quoting zen :)

> Simple is better than complex

shrayasr replied:
> Even if the code is repeated , IMHO it is okay .
> 
> Quoting zen :)
> 
> >    Simple is better than complex

I quite disagree actually. Repetition of code is **bad** in _almost_ all cases. 

Why not have something like an `encryptor_decryptor_core` function that does no such checks of `key <=0`. It is private function that is called by `encrypt` and `decrypt` after making the necessary checks required.

bravegnu replied:
+1 for @shrayasr suggestion.

whiletruelearn replied:
Thanks guys! I will incorporate this change 

what about using another function <code>check_exceptions()</code> where i can put the exception checking logic.

shrayasr replied:
> Thanks guys! I will incorporate this change

:+1: 

> what about using another function check_exceptions() where i can put the exception checking logic.

Quoting from the zen 

> Explicit is better than Implicit 

Let it stay in the public `encrypt()` and `decrypt()` functions. :smile:

whiletruelearn replied:
Hmm :smile: You got me there with zen :sweat_smile: . I would still like to disagree  on the grounds that i don't fully understand what it means. Will catch you today at meetup and get this cleared.

I would update the PR once @bravegnu  confirms on problem with exception.

bravegnu replied:
A function `validate_key()` that checks and throws an error would be nice. And BTW, just a ValueError would suffice. Do you need a separate exception class for this?

shrayasr replied:
> A function validate_key() that checks and throws an error would be nice

I would actually prefer to leave them alone in the individual functions actually. Any reason to move them into another `validate_key` function? The purpose of those checks IMO is better if left with their respective functions. Having a `validate_key` functions means that we'll have to send a key indicating the conditions that the key needs to be checked for.

Eg: 

``` python
def validate_key(fn_type, key):
  if fn_type.strip().upper() == "ENCRYPT":
    # do something
  elif fn_type.strip().upper() == "DECRYPT":
    # do something

def encrypt(...):
  validate_key("encrypt", ...)
  # do something

def decrypt(...):
  validate_key("decrypt", ...)
  # do something
```

Which doesn't look all that elegant to me. 

> And BTW, just a ValueError would suffice. Do you need a separate exception class for this?

+1. Was thinking of this as well

bravegnu replied:
The `validate_key()` will do the same job in both encrypt and decrypt. It will check if the key is a positive integer.

shrayasr replied:
> The `validate_key()` will do the same job in both encrypt and decrypt. It will check if the key is a positive integer.

HMM. I would still leave them in the individual functions.

shrayasr commented:
> @shrayasr Thanks ! Rebase really worked like a charm :sweat_smile:

I wouldn't bank on that always happening :D

whiletruelearn committed:
Removing the custom exceptions and redundant code for encryption and decryption

whiletruelearn commented:
@shrayasr @bravegnu . PR was updated with changes. Could you guys please review.

shrayasr commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,76 @@
+import string
+
+
+def cipher_helper(message, key):
```
Since `cipher_helper` is a _private_ method, it is practice to start them with a double underscore (TIP: pronouced **dunder** or **under under**). Same case with `check_exceptions` as well. 

For more info check out: http://stackoverflow.com/a/1301369/235055

whiletruelearn replied:
so _dunder_ makes it superprivate while single underscore makes it semiprivate

Was wondering the difference between superprivate  and semiprivate in this context.

EDIT : Updated the PR with dunder in front of private methods.

shrayasr replied:
:+1: 

> Was wondering the difference between superprivate and semiprivate in this context.

Not too clear myself. Should read more :)

whiletruelearn committed:
Using dunder for private methods

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,76 @@
+
+def decrypt(message, key):
+    """Returns the actual message which was encrypted
+       with caesar cipher using Key.
```
Make this is a single line. If you want a more detailed description that should be provided in a separate paragraph below the one-line description as shown below.

```
def myfunc(arg1, arg2):
    """"My one-line description.

    Multi-line description is in a separate paragraph, separated from
    the one-line description by a blank line.

    Args:
           ....
    """"
```

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,76 @@
+def encrypt(message, key):
+    """Returns the encrypted message for the string passed as argument.
+
+    Uses a random key for generating the cipher text
```
Not any more.

whiletruelearn replied:
Oops. :smile:

bravegnu commented on libcipher/caesar.py:
```diff
@@ -0,0 +1,76 @@
+__author__ = 'kskrishnasangeeth'
```
A module level docstring would be nice.

bravegnu commented:
@kskrishnasangeeth Looks mostly OK. We can merge this when the docstrings are fixed. Thanks.

whiletruelearn committed:
Fixing docstrings. To Infinity and Beyond

whiletruelearn commented:
@bravegnu Thank You :smiley: I have made the changes.

shrayasr commented:
Other than the small things, LGTM :+1: :ship:

whiletruelearn commented:
@bravegnu @shrayasr 

Is there something else missing or needs to be changed?

bravegnu commented:
Sorry, got held up with work.

whiletruelearn commented:
No problem . I should thank you .

shrayasr commented:
:tada: finally!

whiletruelearn commented:
@shrayasr  Yaay :fireworks:
